server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -1159,6 +1159,32 @@ public <T> Set<T> readSet(Writeable.Reader<T> reader) throws IOException {
         return readCollection(reader, Sets::newHashSetWithExpectedSize, Collections.emptySet());
     }
 
+    /**
+     * Reads a set of objects. The set is expected to have been written using {@link StreamOutput#writeCollection(Collection)}}.
+     * The returned set is immutable.
+     *
+     * @return the set of objects
+     * @throws IOException if an I/O exception occurs reading the set
+     */
+    public <T> Set<T> readImmutableSet(final Writeable.Reader<T> reader) throws IOException {
+        int count = readArraySize();
+        // special cases small arrays, just like in java.util.Set.of(...)
+        if (count == 0) {
```

Comment:
NIT: this could be simplified with a switch expression

Reply:
Hmm not really simpler here, it's far more LoC with a switch I think :)

Reply:
I do not think so:
```
        return switch (count) {
            case 0 -> Set.of();
            case 1 -> Set.of(reader.read(this));
            case 2 -> Set.of(reader.read(this), reader.read(this));
            default -> {
                Object[] entries = new Object[count];
                for (int i = 0; i < count; i++) {
                    entries[i] = reader.read(this);
                }
                @SuppressWarnings("unchecked")
                T[] typedEntries = (T[]) entries;
                yield  Set.of(typedEntries);
            }
        };
```

Reply:
huh fancy :D still the same number of code lines though, looks nice though so I went with this and used it in the duplicate list code also :)